sdk/python/kfp/components/_python_op.py

```diff
@@ -245,16 +249,27 @@ def _func_to_component_spec(func, extra_code='', base_image=_default_base_image,
     arguments = []
     for input in component_spec.inputs:
         param_flag = "--" + input.name.replace("_", "-")
+        is_required = not input.optional and input.default is None
```

Comment:
why isn't 'not optional' sufficient enough for 'is_required'? What are the scenarios when it is required and the input is not optional but the default is not None?

Reply:
>What are the scenarios when it is required and the input is not optional but the default is not None?

Up to this point, `optional` was effectively unused. Many component inputs had default values and that made them effectively optional.
This scheme worked well for most cases, except for the inputs that had default value of `None`. For these inputs it was impossible to distinguish between no default value and `None` default value. Marking those inputs as optional solves the problem.

Technically we could write `is_required = not input.optional` and it would work for most cases. However this will make some scenarios impossible:
* Parameters with complex defaults. It's hard to pass those through the command-line. By marking them as not required, it's possible to not pass anything to the program and make the function use it's complex defaults. Making the argument required will make this impossible:

```
def func(
  param1={'foo', ('bar', 'baz')},
  param2=_singleton,
  param3=_get_default_from_conf('param3')
):
  ...
```
Making the command-line arguments for these parameters not required makes it possible to support such functions. (Actually `None` case can be thought of as a special case of `param2` in the above example).

Reply:
I'm skeptical about this. input_optional does not mean it is not required.
If not used, why not simply is_required=not input.optional.

Reply:
The generated command-line wrapper is an implementation detail of a Lightweight component.
Using `argparse` with `required=False` is an easy and clean solution for the problem of calling python function with complex default values.
When you have a function like that there is nothing you can pass to it that will make it receive its complex default (passing `None` won't work). The only way to make it so that the function receives its default is to not pass anything. And the only way for the command-line interface to indicate the ability to not pass anything is to mark the argument as not required.

Reply:
Why do we now need `optional`? It's needed because there is no other good way to support `None` default values:
```
def func2(a: None):
```
When parsing component.yaml, `{name: a, default: None}` is parsed indistinguishable from `{name: a}`. In both cases we get and instance of the `inputSpec` class with `input.default == None`. Setting `optional: True` allows us to distinguish those cases.

```
def func3(a: "aaa", b: None): 
```
```
inputs:
- {name: a, default: aaa} #implied optional=True
- {name: b, optional: true} #Implied default=None
```

"Optional or default" vs. "optional and default":

The issue is how the parameters with default values are represented in component.yaml:
`def func1(a="aaa"):`
Variant 1:
```
inputs:
- {name: a, default: aaa}
```
Variant 2:
```
inputs:
- {name: a, optional: true, default: aaa}
```
currently all of our component use variant 1. Parameter having a default value implies that it's OK to skip that parameter when calling the function/component. That's why I'm leaning towards having the same for components generated from python - not having to specify both `optional` and `default` at the same time.

Reply:
I've changed it to `is_required = not input.optional` for now, since it's not required for this PR, but I'll add the `.default` check back in next PR.

Note that this changes the code generation for all existing scenarios. I usually try to avoid that if there is no specific need for that.